### PR TITLE
Support to cache data in memory when using Alluxio Local Cache

### PR DESCRIPTION
**Pull request by zzcclp**
### What changes are proposed in this pull request?

Add the third PageStore type: MEM, to support caching data in memory.

### Why are the changes needed?

Please see the description in #14771 .

### Does this PR introduce any user facing changes?

  1. Add a new PageStore type: MEM

### CODE REVIEW

**Comment by zzcclp** [comment]
alluxio-bot, retest this please

**Comment by zzcclp** [comment]
The test case 'alluxio.util.JvmPauseMonitorTest.testMockedInfoPause' failed, I met this  several times.

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+  /**
+   * Creates a new instance of {@link MemoryPageStore}.
+   *
+   * @param options options for the buffer store
+   */
+  public MemoryPageStore(MemoryPageStoreOptions options) {
+    mPageSize = options.getPageSize();
+    mCapacity = (long) (options.getCacheSize() / (1 + options.getOverheadRatio()));
+    mFileBuckets = options.getFileBuckets();
```
Comment (beinan) [inline core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java:52]: do we still need mFileBuckets for mem page store?  I think it's just helpful for file system storage.

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    mPageStoreMap = new ConcurrentHashMap<>();
+  }
+
+  @Override
+  public void put(PageId pageId, byte[] page) throws ResourceExhaustedException, IOException {
+    String pageKey = getKeyFromPageId(pageId);
```
Comment (beinan) [inline core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java:58]: is there any concern to use pageId as the key directly? I didn't see any benefit to use a string key here

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    try {
+      byte[] mPage = new byte[page.length];
+      System.arraycopy(page, 0, mPage, 0, page.length);
```
Comment (beinan) [inline core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java:61]: is the arraycopy here necessary?  I think when we read a page from external file system, we already made a copy there.

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    String pageKey = getKeyFromPageId(pageId);
+    if (!mPageStoreMap.containsKey(pageKey)) {
+      throw new PageNotFoundException(pageKey);
+    }
+    byte[] mPage = mPageStoreMap.get(pageKey);
+    Preconditions.checkArgument(pageOffset <= mPage.length, "page offset %s exceeded page size %s",
+        pageOffset, mPage.length);
+    int bytesLeft = (int) Math.min(mPage.length - pageOffset, buffer.length - bufferOffset);
+    bytesLeft = Math.min(bytesLeft, bytesToRead);
+    System.arraycopy(mPage, pageOffset, buffer, bufferOffset, bytesLeft);
```
Comment (beinan) [inline core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java:85]: arraycopy might make it safer, but is it necessary?

**Review by beinan — COMMENTED** [review]
Great contribution! Our community is waiting for this one for a while. Thank you Zhichao!  it looks good to me except a couple of clarification questions, thanks!

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java (hunk 1 above)
Comment (zzcclp) [reply core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java:52]: I will change the policy for the  key

Comment (zzcclp) [reply core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java]: Already removed mFileBuckets.

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java (hunk 2 above)
Comment (zzcclp) [reply core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java]: Use the 'MD5(file_id(str)_page_idx(ulong)).asLong' as key, the type is long.

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java (hunk 3 above)
Comment (zzcclp) [reply core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java:62]: I think it's necessary, if use the byte array data directly, when the original byte array data is modified, the value is changed too.

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java (hunk 4 above)
Comment (zzcclp) [reply core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java:86]: It's necessary, the reason is the same as the putting method, shown comments above.

**Comment by zzcclp** [comment]
@beinan @apc999 I have modified the code according to the comments, there are some test cases failing, but they are not related to this PR, please help to review, thanks.

**Comment by zzcclp** [comment]
PR is ready, @beinan @apc999 please help to review again, thanks.

**Comment by zzcclp** [comment]
ping @beinan @apc999

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java (hunk 4 above)
Comment (beinan) [reply core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java:86]: but you always make a copy in the put method, I don't think the byte array would be got modified unless there is some other client to modify it (I don't think presto will modify any bytes array it reads from)

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java (hunk 3 above)
Comment (beinan) [reply core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java:62]: hmmm, do you happen to know when and how the original btye array could be modified?  Thanks!

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  }
+
+  /**
+   * @param pageId page Id
+   * @return the key to this page
+   */
+  @VisibleForTesting
+  public long getKeyFromPageId(PageId pageId) {
+    // TODO(feng): encode fileId with URLEncoder to escape invalid characters for file name
+    // Key is : MD5(file_id(str)_page_idx(ulong))
```
Comment (beinan) [inline core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java:107]: You're able to use the pageId direclty as the key, the extra md5 hashing is unnecessary which might cause conflict in some extreme cases.

See the hashcode and equals method in the PageId class

` @Override
  public int hashCode() {
    return Objects.hashCode(mFileId, mPageIndex);
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) {
      return true;
    }
    if (!(obj instanceof PageId)) {
      return false;
    }
    PageId that = (PageId) obj;
    return mFileId.equals(that.mFileId) && mPageIndex == that.mPageIndex;
  }`

Comment (zzcclp) [reply core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java:107]: Got it. I misunderstood your means of comment before.
Done.

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java (hunk 3 above)
Comment (zzcclp) [reply core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java:62]: For example:
```
        InputStream is = new FileInputStream("XXXX/alluxio/NOTICE");
        ByteArrayOutputStream buffer = new ByteArrayOutputStream();

        int nRead;
        long cnt = 0;
        byte[] data = new byte[4];   // read buffer

        while ((nRead = is.read(data, 0, data.length)) != -1) {
            // put data into memory cache
            mPageDataMap.put(cnt, data);
            cnt++;
            buffer.write(data, 0, nRead);
        }

        buffer.flush();
```
Use the byte array as a buffer, read data from a file and put the data into memory cache, at the end, all the cached data is the last one read from the file.

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java (hunk 4 above)
Comment (zzcclp) [reply core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java:86]: > I don't think the byte array would be got modified unless there is some other client to modify it (I don't think presto will modify any bytes array it reads from)

Yeah, like Spark, Presto, they will not modify any data, but we can not guarantee all of the systems will not modify any data after reading from, shown as below:
```
        System.out.println(Arrays.toString(mPageDataMap.get(1l)));
        System.out.println(Arrays.toString(mPageDataMap.get(2l)));

        // read data from memory cache data
        byte[] data1 = mPageDataMap.get(1l);
        // modify the data directly
        data1[0] = 2;
        byte[] data2 = mPageDataMap.get(2l);
        // modify the data directly
        data2[1] = 3;
        
        System.out.println(Arrays.toString(mPageDataMap.get(1l)));
        System.out.println(Arrays.toString(mPageDataMap.get(2l)));
```

the results are:

image

**Comment by zzcclp** [comment]
ping @beinan @apc999

**Comment by zzcclp** [comment]
PR is ready, ping @beinan @apc999, please help to review, thanks.

File: core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java (hunk 4 above)
Comment (beinan) [reply core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java:83]: gotcha, let's keep the array copy for now,  we could improve it if we saw any GC or memory issues.  Also thinking to make the bytes immutable, but it's really difficult to make an immutable array in java.

**Review by beinan — APPROVED** [review]
lgtm,  great contribution!  Thanks a lot!
@apc999 could you help take a look also?

**Comment by beinan** [comment]
alluxio-bot, merge this please

**Pull request merged**